src/ngx_http_lua_balancer.c

```diff
@@ -799,4 +1184,162 @@ ngx_http_lua_ffi_balancer_recreate_request(ngx_http_request_t *r,
+#endif
+    ngx_http_lua_srv_conf_t                 *lscf = conf;
+
+    if (lscf->balancer.max_cached != NGX_CONF_UNSET_UINT) {
+        return "is duplicate";
+    }
+
+    /* read options */
+
+    value = cf->args->elts;
+
+    n = ngx_atoi(value[1].data, value[1].len);
+
+    if (n == NGX_ERROR || n == 0) {
+        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                           "invalid value \"%V\" in \"%V\" directive",
+                           &value[1], &cmd->name);
+        return NGX_CONF_ERROR;
+    }
+
+    lscf->balancer.max_cached = n;
+
+    return NGX_CONF_OK;
+}
+
+
+#if (NGX_HTTP_SSL)
+static ngx_int_t
+ngx_http_lua_upstream_get_ssl_name(ngx_http_request_t *r,
+    ngx_http_upstream_t *u)
+{
+    u_char     *p, *last;
+    ngx_str_t   name;
+
+    if (u->conf->ssl_name) {
+        if (ngx_http_complex_value(r, u->conf->ssl_name, &name) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+    } else {
+        name = u->ssl_name;
+    }
+
+    if (name.len == 0) {
+        goto done;
+    }
+
+    /*
+     * ssl name here may contain port, notably if derived from $proxy_host
+     * or $http_host; we have to strip it
+     */
+
+    p = name.data;
+    last = name.data + name.len;
+
+    if (*p == '[') {
+        p = ngx_strlchr(p, last, ']');
+
+        if (p == NULL) {
+            p = name.data;
+        }
+    }
+
+    p = ngx_strlchr(p, last, ':');
+
+    if (p != NULL) {
+        name.len = p - name.data;
+    }
+
+done:
+
+    u->ssl_name = name;
+
+    return NGX_OK;
+}
+#endif
+
+
+static ngx_uint_t
+ngx_http_lua_balancer_calc_hash(ngx_str_t *name,
+    struct sockaddr *sockaddr, socklen_t socklen)
+{
+    ngx_uint_t hash;
+
+    hash = ngx_hash_key_lc(name->data, name->len);
+    hash ^= ngx_hash_key((u_char *) sockaddr, socklen);
+
+    return hash;
+}
+
+
+static ngx_connection_t *
+ngx_http_lua_balancer_get_cached_item(ngx_http_lua_srv_conf_t *lscf,
+    ngx_peer_connection_t *pc, ngx_str_t *name)
+{
+    ngx_uint_t                         hash;
+    ngx_queue_t                       *q;
+    ngx_queue_t                       *head;
+    ngx_connection_t                  *c;
+    struct sockaddr                   *sockaddr;
+    socklen_t                          socklen;
+    ngx_http_lua_balancer_ka_item_t   *item;
+
+    sockaddr = pc->sockaddr;
+    socklen = pc->socklen;
+
+    hash = ngx_http_lua_balancer_calc_hash(name, sockaddr, socklen);
+    head = &lscf->balancer.buckets[hash % lscf->balancer.bucket_cnt];
+
+    c = NULL;
+    for (q = ngx_queue_head(head);
+        q != ngx_queue_sentinel(head);
+        q = ngx_queue_next(q))
+    {
+        item = ngx_queue_data(q, ngx_http_lua_balancer_ka_item_t, hnode);
+        if (item->hash != hash) {
+            continue;
+        }
+
+        if (ngx_memn2cmp((u_char *) &item->sockaddr,
+                         (u_char *) sockaddr,
+                         item->socklen, socklen) == 0
+            && name->len == item->host.len
```

Comment:
maybe we should move these length compare expressions at the front?
I'm not sure if the C compiler is clever enough to do such optimizations.